src/Mod/TechDraw/App/DrawViewGDTReference.cpp

```diff
@@ -0,0 +1,258 @@
+#include "DrawUtil.h"
+#include "LineGroup.h"
+
+using namespace TechDraw;
+
+//===========================================================================
+// DrawViewGDTReference
+//===========================================================================
+
+PROPERTY_SOURCE(TechDraw::DrawViewGDTReference, TechDraw::DrawView)
+
+const char* DrawViewGDTReference::TypeEnums[]= {"Edge",
+                                         	 "Cosmetic",
+                                             "FeatureFrame",
+											 NULL};
+
+DrawViewGDTReference::DrawViewGDTReference(void)
+{
+	ADD_PROPERTY_TYPE(References2D,(0,0),"",(App::Prop_None),"Projected Geometry References");
+	References2D.setScope(App::LinkScope::Global);
+
+	Type.setEnums(TypeEnums);     // object referenced type : Edge, Cosmetic line, Feature frame.
+	ADD_PROPERTY(Type,((long)0));
+	Type.setStatus(App::Property::ReadOnly,true);
+
+	ADD_PROPERTY_TYPE(Text,("A"),"",App::Prop_None,"The text to be displayed");
+    ADD_PROPERTY_TYPE(SymbolScale,(1),"",(App::PropertyType)(App::Prop_None),"Reference symbol scale");
+
+    m_linearPoints.first  = Base::Vector3d(0,0,0);
+    m_linearPoints.second = Base::Vector3d(0,0,0);
+
+    //hide the DrawView properties that don't apply to reference
+    References2D.setStatus(App::Property::Hidden,true);
+    X.setStatus(App::Property::Hidden,true);
+    Y.setStatus(App::Property::Hidden,true);
+    ScaleType.setStatus(App::Property::ReadOnly,true);
+    ScaleType.setStatus(App::Property::Hidden,true);
+    Scale.setStatus(App::Property::Hidden,true);
+    Rotation.setStatus(App::Property::Hidden,true);
+    Caption.setStatus(App::Property::Hidden,true);
+}
+
+DrawViewGDTReference::~DrawViewGDTReference()
+{
+
+}
+
+//are there non-blank references?
+bool DrawViewGDTReference::has2DReferences(void) const
+{
+    //Base::Console().Message("DrawViewGDTReference::has2DReferences() - %s\n",getNameInDocument());
+    bool result = false;
+
+    const std::vector<App::DocumentObject*> &objects = References2D.getValues();
+    const std::vector<std::string> &SubNames         = References2D.getSubValues();
+    if (!objects.empty()) {
+        App::DocumentObject* testRef = objects.at(0);
+        if (testRef != nullptr) {
+            if (!SubNames.empty()) {
+                result = true;              //not empty is good
+                for (auto& s: SubNames) {   //but check individual entries
+                    if (s.empty()) {
+                        result = false;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    return result;
+}
+
+void DrawViewGDTReference::onChanged(const App::Property* prop)
+{
+    DrawView::onChanged(prop);
+}
+
+void DrawViewGDTReference::onDocumentRestored()
+{
+	execute();
+}
+
+short DrawViewGDTReference::mustExecute() const
+{
+    bool result = 0;
+    if (!isRestoring()) {
+        result = (	References2D.isTouched() ||
+        			Text.isTouched()	);

```

Comment:
Bad spacing. Use spaces, not tabs. Why would you insert a tab before and after parentheses?